src/quickpicks/uris.ts

```diff
@@ -0,0 +1,96 @@
+import {
+  QuickPickItem,
+  QuickPickItemKind,
+  Tab,
+  TabGroup,
+  TabInputText,
+  TextDocument,
+  ThemeIcon,
+  Uri,
+  window,
+  workspace,
+} from "vscode";
+
+/** Create a quickpick to list all open editors, as well as an item to select a file from the open
+ * dialog. */
+export async function uriQuickpick(
+  editorLanguageIds: string[] = [],
+  fileFilters: { [name: string]: string[] } = {},
+): Promise<Uri | undefined> {
+  const selectFileLabel = "Open File...";
+  const quickpickItems: QuickPickItem[] = [
+    {
+      label: selectFileLabel,
+      iconPath: new ThemeIcon("search"),
+      alwaysShow: true,
+    },
+  ];
+
+  const filenameUriMap: Map<string, Uri> = new Map();
+
+  // inspect *all* open editors, not just the visible/active ones
+  const documentPromises: Promise<TextDocument>[] = [];
+  window.tabGroups.all.forEach((tabGroup: TabGroup) => {
+    tabGroup.tabs.forEach(async (tab: Tab) => {
+      // .input is `unknown`, not sure what other properties are available if `uri` isn't there
+      if (!(tab.input instanceof TabInputText)) {
+        return;
+      }
+      const tabInput: TabInputText = tab.input as TabInputText;
+      if (tabInput.uri.scheme === "output") {
+        // ignore output channels
+        return;
+      }
+      // look up document based on Uri
+      documentPromises.push(Promise.resolve(workspace.openTextDocument(tabInput.uri)));
+    });
+  });
+
+  const documents: TextDocument[] = [];
+  const allDocuments: TextDocument[] = await Promise.all(documentPromises);
+  allDocuments.forEach((document: TextDocument) => {
+    if (
+      editorLanguageIds.length === 0 ||
+      (editorLanguageIds.length > 0 && editorLanguageIds.includes(document.languageId))
+    ) {
+      documents.push(document);
+      filenameUriMap.set(document.fileName, document.uri);
+    }
+  });
+
+  let pickedItem: QuickPickItem | undefined;
+  if (documents.length > 0) {
+    quickpickItems.push({
+      kind: QuickPickItemKind.Separator,
+      label: "Open documents",
+    });
+    quickpickItems.push(
+      ...documents.map((document: TextDocument) => {
+        return {
+          label: document.fileName,
+          description: document.languageId,
+          iconPath: new ThemeIcon("file"),
```

Comment:
I haven't found any way to use existing [file icons](https://code.visualstudio.com/api/extension-guides/file-icon-theme) (as opposed to [product icons](https://code.visualstudio.com/api/references/icons-in-labels)) like the built-in "Recent files" quickpick.
<img width="618" alt="image" src="https://github.com/user-attachments/assets/d6026929-a65c-45b8-9e12-3db1565f903f" />